text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+        .spawn()?;
+    // ^ accidentally failed to wait, if programmer wanted to make a daemon
+    //   or zombie or something they should have to write let _ =.
+
+    Command::new("touch")
+        .args(&["/dev/enoent/touch-3"])
+        .spawn()?
+        .wait()?;
+    // ^ accidentally failed to check exit status
+```
+
+Corrected versions tend to have lots of boilerplate code,
+especially if good error messages are wanted.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We will introduce new APIs on `Command`,
+for running the command and collecting its output:
+
+```rust
+impl Command {
+    fn run(&mut self) -> Result<(), SubprocessError>;
+    fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>;
+    fn get_output(&mut self) -> Result<String, SubprocessError>;
+    fn get_output_line(&mut self) -> Result<String, SubprocessError>;
+    fn get_output_read(&mut self) -> impl std::io::Read;
+}
+struct SubprocessError { ... }
+impl From<SubprocessError> for io::Error { ... }
+```
+
+The `.output()` function and `std::process::Output`
+will be deprecated.
+
+No significant changes are made to the `Command` construction APIs,
+but it may become necessary to call `.stderr()` explicitly
+to use the new methods (see Unresolved Questions).
+
+## Use cases
+
+We aim to serve well each of the following people:
+
+ * Alice writes a CLI utility to orchestrate processes and wants top-notch error reporting from subprocesses.
+
+ * Bob migrates ad-hoc automation scripts from bash to Rust.
+
+ * Carol writes a generic application and just wants all errors to be maximally useful by default.
+
+ * Dionysus wants to run `diff`, which exits `0` for "no difference",
+   `1` for "difference found" and
+   another value for failure.
+
+(Partly cribbed from one of
+@matklad's [comments](https://github.com/rust-lang/rust/pull/89004#issuecomment-923803209) in #89004)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## New methods on `Command`:
+
+ * `fn run(&mut self) -> Result<(), SubprocessError>`:
+
+   Runs the command.
+   Equivalent to `.spawn()` followed by `.status()`,
+   but with better error handling.
+
+ * `fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   After the child indicates EOF on its stdout,
+   we will wait for it to finish and check the exit status.
+
+ * `fn get_output(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Does not trim any trailing line ending.
+
+ * `fn get_output_line(&mut self) -> Result<Vec<String>, SubprocessError>`:
+
+   Runs the command and collects its stdout.
+   Decodes the stdout as UTF-8, and fails if that's not possible.
+   Fails unless the output is a single line (with or without line ending).
```

Comment:
Please specify the behavior if the program has *no* output. Does this fail in that case, or is the requirement "at most one line"?

Reply:
I think it should succeed and give an empty string.  Will write that into the RFC.

Reply:
It seems like this helper would commonly be used for programs for which you expect to parse the output, in which case having to separately check for empty string seems to defeat part of the ergonomic win of using the helper.

But in any case, this isn't a blocker as you've split this helper out into possible future work.